File: ExpressionTreeTransform.Tests/CompilerGenerated/Binary.cs
using Xunit;
using static ExpressionTreeTransform.Tests.Runners;
using static ExpressionTreeTransform.Util.Globals;

namespace ExpressionTreeTransform.Tests {
    [Trait("Source", CSharpCompiler)]
    public class Binary {
        [Fact]
        public void Add() {
            double x = 0, y = 0;
            BuildAssert(() => x + y, "() => x + y", "Function() x + y");
        }

        [Fact]
        public void AddChecked() {
            double x = 0, y = 0;
            BuildAssert(() => x + y, "() => x + y", "Function() x + y");
        }

        [Fact]
        public void Divide() {
            double x = 0, y = 0;
            BuildAssert(() => x / y, "() => x / y", "Function() x / y");
        }

        [Fact]
        public void Modulo() {
            double x = 0, y = 0;
            BuildAssert(() => x % y, "() => x % y", "Function() x Mod y");
        }

        [Fact]
        public void Multiply() {
            double x = 0, y = 0;
            BuildAssert(() => x * y, "() => x * y", "Function() x * y");
        }

        [Fact]
        public void MultiplyChecked() {
            double x = 0, y = 0;
            BuildAssert(() => x * y, "() => x * y", "Function() x * y");
        }

        [Fact]
        public void Subtract() {
            double x = 0, y = 0;
            BuildAssert(() => x - y, "() => x - y", "Function() x - y");
        }

        [Fact]
        public void SubtractChecked() {
            double x = 0, y = 0;
            BuildAssert(() => x - y, "() => x - y", "Function() x - y");
        }

        [Fact]
        public void AndBitwise() {
            int i = 0, j = 0;
            BuildAssert(() => i & j, "() => i & j", "Function() i And j");
        }

        [Fact]
        public void OrBitwise() {
            int i = 0, j = 0;
            BuildAssert(() => i | j, "() => i | j", "Function() i Or j");
        }

        [Fact]
        public void ExclusiveOrBitwise() {
            int i = 0, j = 0;
            BuildAssert(() => i ^ j, "() => i ^ j", "Function() i Xor j");
        }

        [Fact]
        public void AndLogical() {
            bool b1 = true, b2 = true;
            BuildAssert(() => b1 & b2, "() => b1 & b2", "Function() b1 And b2");
        }

        [Fact]
        public void OrLogical() {
            bool b1 = true, b2 = true;
            BuildAssert(() => b1 | b2, "() => b1 | b2", "Function() b1 Or b2");
        }

        [Fact]
        public void ExclusiveOrLogical() {
            bool b1 = true, b2 = true;
            BuildAssert(() => b1 ^ b2, "() => b1 ^ b2", "Function() b1 Xor b2");
        }

        [Fact]
        public void AndAlso() {
            bool b1 = true, b2 = true;
            BuildAssert(() => b1 && b2, "() => b1 && b2", "Function() b1 AndAlso b2");
        }

        [Fact]
        public void OrElse() {
            bool b1 = true, b2 = true;
            BuildAssert(() => b1 || b2, "() => b1 || b2", "Function() b1 OrElse b2");
        }

        [Fact]
        public void Equal() {
            int i = 0, j = 0;
            BuildAssert(() => i == j, "() => i == j", "Function() i = j");
        }

        [Fact]
        public void NotEqual() {
            int i = 0, j = 0;
            BuildAssert(() => i != j, "() => i != j", "Function() i <> j");
        }

        [Fact]
        public void GreaterThanOrEqual() {
            int i = 0, j = 0;
            BuildAssert(() => i >= j, "() => i >= j", "Function() i >= j");
        }

        [Fact]
        public void GreaterThan() {
            int i = 0, j = 0;
            BuildAssert(() => i > j, "() => i > j", "Function() i > j");
        }

        [Fact]
        public void LessThan() {
            int i = 0, j = 0;
            BuildAssert(() => i < j, "() => i < j", "Function() i < j");
        }

        [Fact]
        public void LessThanOrEqual() {
            int i = 0, j = 0;
            BuildAssert(() => i <= j, "() => i <= j", "Function() i <= j");
        }

        [Fact]
        public void Coalesce() {
            string s1 = null, s2 = null;
            BuildAssert(() => s1 ?? s2, "() => s1 ?? s2", "Function() If(s1, s2)");
        }

        [Fact]
        public void LeftShift() {
            int i = 0, j = 0;
            BuildAssert(() => i << j, "() => i << j", "Function() i << j");
        }

        [Fact]
        public void RightShift() {
            int i = 0, j = 0;
            BuildAssert(() => i >> j, "() => i >> j", "Function() i >> j");
        }

        [Fact]
        public void ArrayIndex() {
            var arr = new string[] { };
            BuildAssert(() => arr[0], "() => arr[0]", "Function() arr(0)");
        }
    }
}

File: ExpressionTreeTransform.Tests.VB/CompilerGenerated/Binary.vb
Public Class Binary
    <Fact>
    Public Sub Add()
        Dim x As Double = 0, y As Double = 0
        BuildAssert(Function() x + y, "() => x + y", "Function() x + y")
    End Sub

    <Fact>
    Public Sub AddChecked()
        Dim x As Double = 0, y As Double = 0
        BuildAssert(Function() x + y, "() => x + y", "Function() x + y")
    End Sub

    <Fact>
    Public Sub Divide()
        Dim x As Double = 0, y As Double = 0
        BuildAssert(Function() x / y, "() => x / y", "Function() x / y")
    End Sub

    <Fact>
    Public Sub Modulo()
        Dim x As Double = 0, y As Double = 0
        BuildAssert(Function() x Mod y, "() => x % y", "Function() x Mod y")
    End Sub

    <Fact>
    Public Sub Multiply()
        Dim x As Double = 0, y As Double = 0
        BuildAssert(Function() x * y, "() => x * y", "Function() x * y")
    End Sub

    <Fact>
    Public Sub MultiplyChecked()
        Dim x As Double = 0, y As Double = 0
        BuildAssert(Function() x * y, "() => x * y", "Function() x * y")
    End Sub

    <Fact>
    Public Sub Subtract()
        Dim x As Double = 0, y As Double = 0
        BuildAssert(Function() x - y, "() => x - y", "Function() x - y")
    End Sub

    <Fact>
    Public Sub SubtractChecked()
        Dim x As Double = 0, y As Double = 0
        BuildAssert(Function() x - y, "() => x - y", "Function() x - y")
    End Sub

    <Fact>
    Public Sub AndBitwise()
        Dim i As Integer = 0, j As Integer = 0
        BuildAssert(Function() i And j, "() => i & j", "Function() i And j")
    End Sub

    <Fact>
    Public Sub OrBitwise()
        Dim i As Integer = 0, j As Integer = 0
        BuildAssert(Function() i Or j, "() => i | j", "Function() i Or j")
    End Sub

    <Fact>
    Public Sub ExclusiveOrBitwise()
        Dim i As Integer = 0, j As Integer = 0
        BuildAssert(Function() i Xor j, "() => i ^ j", "Function() i Xor j")
    End Sub

    <Fact>
    Public Sub AndLogical()
        Dim b1 As Boolean = True, b2 As Boolean = True
        BuildAssert(Function() b1 And b2, "() => b1 & b2", "Function() b1 And b2")
    End Sub

    <Fact>
    Public Sub OrLogical()
        Dim b1 As Boolean = True, b2 As Boolean = True
        BuildAssert(Function() b1 Or b2, "() => b1 | b2", "Function() b1 Or b2")
    End Sub

    <Fact>
    Public Sub ExclusiveOrLogical()
        Dim b1 As Boolean = True, b2 As Boolean = True
        BuildAssert(Function() b1 Xor b2, "() => b1 ^ b2", "Function() b1 Xor b2")
    End Sub

    <Fact>
    Public Sub [AndAlso]()
        Dim b1 As Boolean = True, b2 As Boolean = True
        BuildAssert(Function() b1 AndAlso b2, "() => b1 && b2", "Function() b1 AndAlso b2")
    End Sub

    <Fact>
    Public Sub [OrElse]()
        Dim b1 As Boolean = True, b2 As Boolean = True
        BuildAssert(Function() b1 OrElse b2, "() => b1 || b2", "Function() b1 OrElse b2")
    End Sub

    <Fact>
    Public Sub Equal()
        Dim i As Integer = 0, j As Integer = 0
        BuildAssert(Function() i = j, "() => i == j", "Function() i = j")
    End Sub

    <Fact>
    Public Sub NotEqual()
        Dim i As Integer = 0, j As Integer = 0
        BuildAssert(Function() i <> j, "() => i != j", "Function() i <> j")
    End Sub

    <Fact>
    Public Sub GreaterThanOrEqual()
        Dim i As Integer = 0, j As Integer = 0
        BuildAssert(Function() i >= j, "() => i >= j", "Function() i >= j")
    End Sub

    <Fact>
    Public Sub GreaterThan()
        Dim i As Integer = 0, j As Integer = 0
        BuildAssert(Function() i > j, "() => i > j", "Function() i > j")
    End Sub

    <Fact>
    Public Sub LessThan()
        Dim i As Integer = 0, j As Integer = 0
        BuildAssert(Function() i < j, "() => i < j", "Function() i < j")
    End Sub

    <Fact>
    Public Sub LessThanOrEqual()
        Dim i As Integer = 0, j As Integer = 0
        BuildAssert(Function() i <= j, "() => i <= j", "Function() i <= j")
    End Sub

    <Fact>
    Public Sub Coalesce()
        Dim s1 As String = Nothing, s2 As String = Nothing
        BuildAssert(Function() If(s1, s2), "() => s1 ?? s2", "Function() If(s1, s2)")
    End Sub

    <Fact>
    Public Sub LeftShift()
        Dim i As Integer = 0, j As Integer = 0
        BuildAssert(Function() i << j, "() => i << j", "Function() i << j")
    End Sub

    <Fact>
    Public Sub RightShift()
        Dim i As Integer = 0, j As Integer = 0
        BuildAssert(Function() i >> j, "() => i >> j", "Function() i >> j")
    End Sub

    <Fact> Public Sub ArrayIndex()
        Dim arr = New String() {}
        BuildAssert(Function() arr(0), "() => arr[0]", "Function() arr(0)")
    End Sub

    <Fact> Public Sub Power()
        Dim x As Double = 0, y As Double = 0
        BuildAssert(Function() x ^ y, "() => Math.Pow(x, y)", "Function() x ^ y")
    End Sub
End Class
